git failure should not break build if ubench not requested

By default cmake will fail if the submodule update call to git fails, even if the microbenchmarks have not been requested. It should continue.

This is on a system where git is found, however the executable has been compiled for a different architecture to the one on which we are building (x86 vs. Power8)

```
bcumming@openpower01:build > cmake ..
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /cm/shared/apps/openpower/software/GCC/5.4.0/bin/gcc
-- Check for working C compiler: /cm/shared/apps/openpower/software/GCC/5.4.0/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /cm/shared/apps/openpower/software/GCC/5.4.0/bin/g++
-- Check for working CXX compiler: /cm/shared/apps/openpower/software/GCC/5.4.0/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Could NOT find UNWIND (missing:  UNWIND_INCLUDE_DIR UNWIND_LIBRARIES) 
-- Found Git: /users/bcumming/install/bin/git  
running cd "/users/bcumming/nestmc-proto/tests/ubench" && /users/bcumming/install/bin/git submodule update --init google-benchmark  2>&1
/bin/sh: /users/bcumming/install/bin/git: cannot execute binary file
CMake Error at /cm/shared/apps/openpower/software/CMake/3.7.1/share/cmake-3.7/Modules/ExternalProject.cmake:1976 (message):
  No download info given for 'gbench' and its source directory:

   /users/bcumming/nestmc-proto/tests/ubench/google-benchmark

  is not an existing non-empty directory.  Please specify one of:

   * SOURCE_DIR with an existing non-empty directory
   * URL
   * GIT_REPOSITORY
   * HG_REPOSITORY
   * CVS_REPOSITORY and CVS_MODULE
   * SVN_REVISION
   * DOWNLOAD_COMMAND
Call Stack (most recent call first):
  /cm/shared/apps/openpower/software/CMake/3.7.1/share/cmake-3.7/Modules/ExternalProject.cmake:2525 (_ep_add_download_command)
  tests/ubench/CMakeLists.txt:31 (ExternalProject_Add)


-- Configuring incomplete, errors occurred!
See also "/users/bcumming/nestmc-proto/build/CMakeFiles/CMakeOutput.log".
See also "/users/bcumming/nestmc-proto/build/CMakeFiles/CMakeError.log".
```

--- a/tests/ubench/CMakeLists.txt
+++ b/tests/ubench/CMakeLists.txt
@@ -20,15 +20,30 @@ set(gbench_cmake_args
 # Attempt to update git submodule if required.
 find_package(Git)
 if(NOT EXISTS "${gbench_src_dir}/.git")
+    set(git_failed)
     if(GIT_FOUND)
-        exec_program("${GIT_EXECUTABLE}" "${CMAKE_CURRENT_SOURCE_DIR}"
-            ARGS submodule update --init google-benchmark)
+        message(STATUS "Updating the google-benchmark submodule ${gbench_src_dir}")
+        execute_process(
+            COMMAND "${GIT_EXECUTABLE}" submodule update --init "${gbench_src_dir}"
+            WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+            ERROR_VARIABLE git_error
+            RESULT_VARIABLE git_result)
+        if(NOT git_result EQUAL 0)
+            set(git_failed "${git_error}")
+        endif()
     else()
-        message(WARNING "Unable to update the google-benchmark submodule: git not found.")
+        set(git_failed "git not found")
+    endif()
+
+    if(git_failed)
+        message(WARNING "Unable to update the google-benchmark submodule: ${git_failed}")
     endif()
 endif()
 
 ExternalProject_Add(gbench
+    # Add dummy DOWNLOAD_COMMAND to stop ExternalProject_Add terminating CMake if the
+    # git submodule had not been udpated.
+    DOWNLOAD_COMMAND "${CMAKE_COMMAND}" -E echo "Warning: ${gbench_src_dir} empty or missing."
     SOURCE_DIR "${gbench_src_dir}"
     CMAKE_ARGS "${gbench_cmake_args}"
     INSTALL_DIR "${gbench_install_dir}"

